include/arch/arm/arch/machine/gic_v3.h

```diff
@@ -291,3 +347,195 @@ static inline void ackInterrupt(irq_t irq)
 
 }
 
+#ifdef CONFIG_ARM_HYPERVISOR_SUPPORT
+
+extern unsigned int gic_vcpu_num_list_regs;
+
+static inline uint32_t get_gic_vcpu_ctrl_hcr(void)
+{
+    uint32_t reg;
+    MRS(ICH_HCR_EL2, reg);
+    return reg;
+}
+
+static inline void set_gic_vcpu_ctrl_hcr(uint32_t hcr)
+{
+    MSR(ICH_HCR_EL2, hcr);
+}
+
+static inline uint32_t get_gic_vcpu_ctrl_vmcr(void)
+{
+    uint32_t reg;
+    MRS(ICH_VMCR_EL2, reg);
+    return reg;
+}
+
+static inline void set_gic_vcpu_ctrl_vmcr(uint32_t vmcr)
+{
+    MSR(ICH_VMCR_EL2, vmcr);
+}
+
+/* Note: On the GICv3 there are potentially up to 128 preemption
+ * levels, and as such up to 4 APR registers.
+ *
+ * At this point the code assumes a maximum of 32 preemption levels.
```

Comment:
I think this can work because we leave the max priority for decodeVCPUInjectIRQ as 31.